File: Microsoft/product_array_except_self/ProductArrayExceptSelf.java
int product[] = new int[nums.length];
product[0] = 1;
int sp = 1;

for (int i = 1; i < nums.length; i++) {
    product[i] = nums[i - 1] * product[i - 1];
}
for (int i = nums.length - 1; i >= 0; i--) {
    product[i] = sp * product[i];
    sp = sp * nums[i];
}

return product;

File: Microsoft/product_array_except_self/product_array_except_self.cpp
class ProblemSolve {
    public:
        vector<int> productExceptSelf(vector<int>& nums) {
            int product = 1;
            int f = 0;
            for (int i = 0; i < nums.size(); i++) {
                product *= nums[i];
                if (nums[i] == 0)
                    f++;
            }

            if (f == nums.size())
                return nums;
            
            int productWoZero = 1;
            if (f > 1)
                productWoZero = 0;
            else {
                for (int i = 0; i < nums.size(); i++) {
                    if (nums[i] != 0)
                        productWoZero *= nums[i];
                }
            }

            for (int i = 0; i < nums.size(); i++) {
                if (nums[i] == 0)
                    nums[i] = productWoZero;
                else
                    nums[i] = product / nums[i];
            }

            return nums;
        }
};
